### PR TITLE
Name inference rule for torch.cat

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25569 Name inference rules for relu/relu_/threshold/threshold_
* **#25568 Name inference rule for torch.cat**
* #25567 Name inference for masked_fill_ / masked_fill
* #25566 Name inference rule for masked select

Test Plan
- new test [namedtensor ci]

Differential Revision: [D17159069](https://our.internmc.facebook.com/intern/diff/D17159069)

### CODE REVIEW

File: aten/src/ATen/NamedTensorUtils.cpp [hunk 1]
```diff
@@ -51,33 +51,38 @@ static void report_positional_error(
     const Dimname& name,
     const Dimname& other_name,
     DimnameList names,
-    DimnameList other_names) {
+    DimnameList other_names,
+    const char* action) {
   // TODO(zou3519): Can improve message by checking if names are alignable and suggesting workarounds
   TORCH_CHECK(false,
-      "Error when attempting to broadcast dims ", names, " and dims ",
+      "Error when attempting to ", action, " dims ", names, " and dims ",
```
Comment (nairbv) [inline aten/src/ATen/NamedTensorUtils.cpp]: It might be worth truncating the string for `action` if someone has many names, though it might be rare.

Comment (zou3519) [reply aten/src/ATen/NamedTensorUtils.cpp]: > It might be worth truncating the string for action if someone has many names, though it might be rare.

What do you mean by "truncating the string for action"?

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -50,8 +50,30 @@ static void check_cat_no_zero_dim(TensorList tensors) {
...
+  }
+  namedinference::propagate_names(result, std::move(outnames), /*validate_names=*/false);
+#endif
+  return result;
+}
+
+#ifdef BUILD_NAMEDTENSOR
+Tensor& cat_out(Tensor& result, TensorList tensors, Dimname dim) {
+  TORCH_CHECK(!tensors.empty(), "expected a non-empty list of Tensors");
+  return at::cat_out(result, tensors, dimname_to_position(tensors[0], dim));
```
Comment (nairbv) [inline aten/src/ATen/native/TensorShape.cpp]: Not sure if this is correct but I'm thinking that if the passed in `dim` refers to a specific tag, and tensors[0]'s dimnames are tagged differently than tensors[1]'s (e.g. A.x, A.y vs A.y, A.x), then should that be an error? 

I think in this case if `dim==A.x` it would find the first dimension in the first tensor, and A.x would be compatible with A.y, but since the tag was explicit maybe shouldn't be? or maybe I'm wrong and A.x is already incompatible with A.y? Regardless a test with tags might be helpful to clarify what's expected.

Comment (zou3519) [reply aten/src/ATen/native/TensorShape.cpp]: We're not going to include tagged names in the first iteration of named tensors. Eventually, before the release, I will send a PR to disable their creation.

For context: I've been thinking about tagged names a lot and I think that they are most likely a bad idea and might not exist in a future iteration of named tensors; as you pointed out, there are a lot of edge cases with tags.

File: aten/src/ATen/NamedTensorUtils.cpp (hunk 1 above)
Comment (nairbv) [reply aten/src/ATen/NamedTensorUtils.cpp]: sorry I meant `names`

File: aten/src/ATen/native/TensorShape.cpp (hunk 2 above)
Comment (nairbv) [reply aten/src/ATen/native/TensorShape.cpp]: ah OK, would like to see more on the thinking there and what we're going to do. It seems like a meaningful use case and we do already have a lot of code for handling them, but yes it does seem pretty complicated.

**Review by nairbv — APPROVED** [review]
